lib/aws-record/record/batch_read.rb

```diff
@@ -0,0 +1,154 @@
+          @unprocessed_keys = unprocessed_keys[BATCH_GET_ITEM_LIMIT..-1].dup
+        else
+          operation_keys = unprocessed_keys.dup
+          @unprocessed_keys.clear
+        end
+
+        operations = build_operations(operation_keys)
+        result = @client.batch_get_item(request_items: operations)
+        build_items(result.responses)
+
+        unless result.unprocessed_keys.nil?
+          update_unprocessed_keys(result.unprocessed_keys)
+        end
+
+        self
+      end
+
+      def complete?
+        unprocessed_keys.none?
+      end
+
+      def items
+        @items ||= []
+      end
+
+      private
+      def unprocessed_keys
+        @unprocessed_keys ||=[]
+      end
+
+      def item_classes
+        @item_classes ||= {}
+      end
+
+      def format_unprocessed_key(klass, key)
+        item_key = {}
+        attributes = klass.attributes
+        klass.keys.each_value do |attr_sym|
+          unless key[attr_sym]
+            raise Errors::KeyMissing.new(
+              "Missing required key #{attr_sym} in #{key}"
+            )
+          end
+          attr_name = attributes.storage_name_for(attr_sym)
+          item_key[attr_name] = attributes.attribute_for(attr_sym).
+            serialize(key[attr_sym])
+        end
+        item_key
+      end
+
+      def store_unprocessed_key(klass, unprocessed_key)
+        unprocessed_keys << {:keys => unprocessed_key, :table_name => klass.table_name}
+      end
+
+      def store_item_class(klass, key)
+        if item_classes.include?(klass.table_name)
+          item_classes[klass.table_name].each do |item|
+            if item[:keys] == key && item[:class] != klass
+              raise 'Provided item keys is a duplicate request'
+            end
+          end
+        end
+        item_classes[klass.table_name] ||= []
+        item_classes[klass.table_name] << {keys: key, class: klass}
+      end
+
+      def build_operations(keys)
+        operations = {}
+        keys.each do | key |
+          operations[key[:table_name]] ||= { keys: [] }
+          operations[key[:table_name]][:keys] << key[:keys]
+        end
+        operations
+      end
+
+      def build_items(item_responses)
+        item_responses.each do |table, unprocessed_items|
+          unprocessed_items.each do |item|
+            item_class = find_item_class(table, item)
+            if item_class.nil?
+              raise 'Item Class was not found'
+            end
+            item = build_item(item, item_class)
+            items << item
+          end
+        end
+      end
+
+      def update_unprocessed_keys(keys)
+        keys.each do |table_name, table_values|
+          table_values.keys.each do |key|
+            formatted_key = {:keys => key, :table_name => table_name}
+            unprocessed_keys << formatted_key
+          end
+        end
+      end
+
+      def find_item_class(table, item)
+        item_class = nil
+        item_classes[table].find do |item_info|
```

Comment:
Instead of using .find with a break we could instead use `select.first` (methods like `select` return an enumerator, it won't apply the block to every item when you do `first` after.  And if no items are found, it will return nil safely.  So something like:
```
item_classes[table].select { |item_info| contains_keys?(item, item_info[:keys]) }.first
```

Reply:
Good to know, updated!

Reply:
`find` is better here (see my comment)